server/sonar-server/src/test/java/org/sonar/server/usergroups/ws/SearchActionTest.java

```diff
@@ -0,0 +1,176 @@
+
+  @ClassRule
+  public static final DbTester dbTester = new DbTester();
+
+  WebService.Controller controller;
+
+  WsTester tester;
+
+  private GroupDao groupDao;
+
+  private GroupMembershipDao groupMembershipDao;
+
+  private UserGroupDao userGroupDao;
+
+  private DbSession session;
+
+  @Before
+  public void setUp() {
+    dbTester.truncateTables();
+
+    groupDao = new GroupDao(mock(System2.class));
+    groupMembershipDao = new GroupMembershipDao(dbTester.myBatis());
+    userGroupDao = new UserGroupDao();
+
+    DbClient dbClient = new DbClient(dbTester.database(), dbTester.myBatis(), groupDao, groupMembershipDao);
+
+    tester = new WsTester(new UserGroupsWs(new SearchAction(dbClient)));
+    controller = tester.controller("api/usergroups");
+
+    session = dbClient.openSession(false);
+  }
+
+  @After
+  public void after() {
+    session.close();
+  }
+
+  @Test
+  public void search_empty() throws Exception {
+    tester.newGetRequest("api/usergroups", "search").execute().assertJson(getClass(), "empty.json");
+  }
+
+  @Test
+  public void search_without_parameters() throws Exception {
+    insertGroups("users", "admins", "customer1", "customer2", "customer3");
+    session.commit();
+
+    tester.newGetRequest("api/usergroups", "search").execute().assertJson(getClass(), "five_groups.json");
+  }
+
+  @Test
+  public void search_with_members() throws Exception {
+    insertGroups("users", "admins", "customer1", "customer2", "customer3");
+    insertMembers("users", 5);
+    insertMembers("admins", 1);
+    insertMembers("customer2", 4);
+    session.commit();
+
+    tester.newGetRequest("api/usergroups", "search").execute().assertJson(getClass(), "with_members.json");
+  }
+
+  @Test
+  public void search_with_query() throws Exception {
+    insertGroups("users", "admins", "customer1", "customer2", "customer3");
+    session.commit();
+
+    tester.newGetRequest("api/usergroups", "search").setParam(WebService.Param.TEXT_QUERY, "custom").execute().assertJson(getClass(), "customers.json");
+  }
+
+  @Test
+  public void search_with_paging() throws Exception {
+    insertGroups("users", "admins", "customer1", "customer2", "customer3");
+    session.commit();
+
+    tester.newGetRequest("api/usergroups", "search")
+      .setParam(WebService.Param.PAGE_SIZE, "3").execute().assertJson(getClass(), "page_1.json");
+    tester.newGetRequest("api/usergroups", "search")
+      .setParam(WebService.Param.PAGE_SIZE, "3").setParam(WebService.Param.PAGE, "2").execute().assertJson(getClass(), "page_2.json");
+    tester.newGetRequest("api/usergroups", "search")
+      .setParam(WebService.Param.PAGE_SIZE, "3").setParam(WebService.Param.PAGE, "3").execute().assertJson(getClass(), "page_3.json");
+  }
+
+  @Test
+  public void search_with_fields() throws Exception {
+    insertGroups("sonar-users");
+    session.commit();
+
+    assertThat(tester.newGetRequest("api/usergroups", "search").execute().outputAsString())
```

Comment:
I prefer to keep DAO tests separate from WS tests:
- coverage from WS tests is not counted for DAO in sonar-core
- when WS implementation changes, DAO is still tested